### ISSUE TITLE
Notes are irregularly corrupted when changing in edit mode

### ISSUE THREAD

**Issue by mikaelmello** [notable/notable#574]
- **OS Version**: Ubuntu 16.04
- **Notable Version**: v1.4.0

### Current behavior

When in edit mode, if your current note has an unsaved changed title (1), when you click on another note (2), it might overwrite Note 1 with the contents of Note 2.

It happened yesterday and now today. After multiple tries I managed to reproduce by quickly adding a lot of text (pasting lorem ipsum multiple times) and clicking in the other note. It rarely happens though, making it very hard to reproduce.

### Expected behavior

Not lose note's contents. I'm very sad to lose important work twice now.

This has actually happened three times:
1. Big essay with a project proposal, it got corrupted and untagged, I was able to recover it from Google Drive
2. 2 hours worth of notes from an important class. At this time I thought it was a problem with syncing and bad connection, decided to only use local files
3. ~1 hour worth of notes from another class.

Unfortunately I'm going to have to halt my use of Notable because I'm afraid of losing even more work.

**Comment by mikaelmello** [comment]
I've realized that whenever I click on a note 2 when there's an edit in the title of current Note 1, it saves the current note and I have to click again to change notes. 

And in the cases where the bug happened, it didn't require 2 clicks, it went straight into note 2.

I'm pretty sure that's related to the file management, just a guess:

- Editing note 1 for a long time, lots of buffered changes
- Click on note 2, sometimes it does nothing but save note 1, but when the bug happens, it changes the internal state of current tab
- We are still saving Note 1, but Notable's state has data from note 2, the buffer is probably overwritten and we save note 2's contents into Note 1

**Comment by mikaelmello** [comment]
Adding to my guess: it's probably a race condition since it happens so irregularly, maybe the changes added to wait for FS events did not cover this condition

**Comment by mspoehr** [comment]
I've had something similar to this happen where one note will overwrite the contents of another. I assumed that the bug fixes to file management in v1.3.0 would resolve this, but apparently not. It sounds similar to #425 and #459.

In my case, I was able to recover the lost file because my data directory is inside of Dropbox, but I wanted to affirm that this has effected multiple people.

**Comment by mikaelmello** [comment]
> In my case, I was able to recover the lost file because my data directory is inside of Dropbox, but I wanted to affirm that this has effected multiple people.

That was my case too until I decided to stop syncing, bad idea. Well, maybe the corruption would be synced too. I guess I'll just add a cron job to commit my changes every minute or so or use VSCode to manually edit everything (they have a great markdown preview too)

**Comment by fabiospampinato** [comment]
First of all I'm very sorry to hear about this, clearly the codebase needs to be more robust and some comprehensive tests must be written especially for guarding against data-loss issues. I'll try to get this major reorganization/testing done by the end of next month, but it will require some considerable amount of time to accomplish this, maybe it will take more than 1 month.

> Adding to my guess: it's probably a race condition since it happens so irregularly, maybe the changes added to wait for FS events did not cover this condition

That would be my guess too, I haven't figured out how this happens yet though.

>  I assumed that the bug fixes to file management in v1.3.0 would resolve this, but apparently not. 

This sounds like a bug with similar results but a different cause probably.

> I guess I'll just add a cron job to commit my changes every minute or so

I supposed I should probably prioritize native version control #116 support more, as that could also be a last-resort measure against data-loss problems. But that's a tricky problem to tackle, as in the not so distant future we will want to enable synchronization & version control from the web app too, git & dropbox/gdrive/etc. don't play too well with each other, and maybe git isn't the best tool if at some point in the future we'll want to add some collaborative editing features. I haven't figured out yet what the best tool for the job is.

---

Getting back to the main issue I haven't been able to reproduce this in its entirety yet: If I change a note's title and then click on the notes list then the list will refresh and I will have to click the list again to switch note, this part I can reproduce (and it seems super annoying), but I don't see the data loss happening.

> And in the cases where the bug happened, it didn't require 2 clicks, it went straight into note 2.

That's interesting, I'll investigate this more. 

Hopefully we'll get a patch release addressing this issue later today or tomorrow.

**Comment by jasonwhat** [comment]
Is this only happening when using something to sync? I'm using Google Drive mounted to linux on ChromeOS and I've been able to go into the version control of GDrive and see file names switching and overwriting existing files. Luckily I'd backed up my one long note elsewhere before it was overwritten.

**Comment by mikaelmello** [comment]
> Is this only happening when using something to sync? 

No, it happened on local files too, although very rarely.

**Comment by jasonwhat** [comment]
Is a fix coming soon or is it recommended to go back to 1.3?

**Comment by mikaelmello** [comment]
@jasonwhat, I don't think there's an ETA about the fix. But I also do not recommend to go back to 1.3. This issue happens more frequently on v1.3 than on v1.4. It only happens very rarely on v1.4

**Comment by mikaelmello** [comment]
I've edited the title in order to not lead anyone into thinking this was introduced in v1.4

**Comment by fabiospampinato** [comment]
@jasonwhat as @mikaelmello said I also wouldn't recommend switching back to v1.3. There's no specific ETA, but this issue has the highest priority in my list of issues to work on [next](https://github.com/notable/notable/milestone/1). Also I've recently [stated](https://github.com/notable/notable/issues/545) that I'd like to target the 1st of each month as a release date, with an exception for issues that cause data loss, but unfortunately I've been super busy in real life in the past few days and I haven't had much time to work on Notable, but while I haven't made any significant commits since the last release I'm actually also rewriting our state management library from scratch, as I think that's currently the single biggest roadblock for us with regards to performance.

TL;DR: I've been busy and I haven't been able to reproduce this issue yet, but time allowing we should have a new release fixing this specific issue within the next few days, apologies for the delay.

**Comment by lmihalkovic** [comment]
@fabiospampinato ... now perhaps you understand differently what i said about the inherent fragility of the code

**Comment by mikaelmello** [comment]
@lmihalkovic Could you please elaborate on that and what is the fragile aspect of the current architecture?

Asking this out of sheer curiosity in order to learn and not make the same (possible ?) mistakes in my own projects.

**Comment by fabiospampinato** [comment]
@lmihalkovic I can't say that I do, as I didn't think the codebase was perfect to begin with.

@mikaelmello I think @lmihalkovic refers to the fact that the various components that comprise this app aren't well-enough separated, ideally you would want to design self-contained components so that you can reason about them pretty much individually rather than having also to think about how the may interact with each other and how changing one thing in a component may break something in another, not strictly related, component.

**Comment by lmihalkovic** [comment]
@mikaelmello our host stated that life was too short to do so, but I found that reading what Joel S. on leaky abstractions was actually a time saver .. 😉 .. peace to both

**Comment by gramster** [comment]
This just happened to me for the third time too.

**Comment by dvergeylen** [comment]
@gramster, I experienced the same issue twice, but couldn't retrieve the files (I deleted the Tutorial notebook) sadly. :disappointed:

**Comment by gramster** [comment]
Did you check your notes folder? Perhaps it ended up in some other random notebook? I think I was looking at the tutorial just before as I wanted to link notes. So perhaps it ended up in a recently viewed notebook.

**Comment by Armagedoom** [comment]
If this happens to you again, and until the issue is solved, seems that a good practice would be to go to the folder containing the notes and sort them in recently edited order (within the file information, not the note metadata, just a file system edit date sort). 
If they are not being erased, they should be easy to spot.

**Comment by fabiospampinato** [comment]
I'm now able to reproduce this reliably with the following script (assuming there's a note with "Test" in its title and another one with "Control" in its):

```js

function attempt () {
  var control = UNSTATED.containers.Notes.state.notes[Object.keys ( UNSTATED.containers.Notes.state.notes ).find ( p => p.includes ( 'Control' ) )];
  var test = UNSTATED.containers.Notes.state.notes[Object.keys ( UNSTATED.containers.Notes.state.notes ).find ( p => p.includes ( 'Test' ) )];
  setTimeout ( () => {
    UNSTATED.containers.Editor.toggleEditing ( true );
    UNSTATED.containers.Note.set ( test, true );
    setTimeout ( () => {
      UNSTATED.containers.Editor.state.monaco.executeEdits ( '', [
        { range: new monaco.Range ( 0, 0, 0, Infinity ), text: `# Test: ${Date.now ()}\n` }
      ]);
      setTimeout ( () => {
        UNSTATED.containers.Note.set ( control, true );
      }, 60 );
    }, 30 );
  }, 5 );  
}

setInterval ( attempt, 100 );
```

My best guess would be that there's some kind of race condition, where the the UI gets updated before the current note had a chance of being saved, and when it comes the time to save the note the app thinks we are actually editing the next note we just switched to. 

The short-term fix is probably tying the editor and the note it is currently editing more together (this might be tricky in some situations, e.g. maybe the note you're currently editing gets renamed on disk, this is another issue that we have to deal with because we don't have UUIDs #538 (even UUIDs stored in the metadata section of the notes would be enough)), but without a reliable way of reproducing the bug it was difficult to be actually sure that the problem got fixed.

In the long-ish-term the backend will have to be rewritten, currently there are too many components interacting with each other and that makes it difficult to fix these kinds of bugs.

**Comment by lmihalkovic** [comment]
... hmm ... reading that last paragraph, __told you so__ comes to mind

**Comment by fabiospampinato** [comment]
@lmihalkovic I don't think I've ever said that the whole thing was perfectly written in the first place, if that's even possible. Please keep these sorts of comments to a minimum, as I don't think they are helping anybody.

**Comment by fabiospampinato** [comment]
I think I've probably just fixed this [here](https://github.com/notable/notable/commit/80c8ebf233fa8cac954f4bc2a2671d58ee129080), but the method I used to reproduce this bug might have been itself buggy, so later today/tomorrow I'm going to attach a test release here, so that you guys that can reproduce this more reliably than me can check if the bug still happens before the new version is released, that would help a lot.

**Comment by lmihalkovic** [comment]
The comment was meant for you.. and its helpfulness is a direct function of who you chose to be. Some feedback was shared with you. this could have prompted you to try and figure what someone could have seen that prompted it... but instead you chose a path of waiting for users to point you at the symptoms of actual problems in order to fix them specifically. And even now, you are starting to touch the issues.. but you make the decision to delay fixing them.. pushing onto users the responsibility to point out the next problem... you do it with a graceful smile, but you do it nonetheless.... no problem..

**Comment by fabiospampinato** [comment]
Sorry @lmihalkovic, I didn't want to resort to this but I just blocked you from this organization, I have no time for dealing with this sorts of comments. Best of luck with everything.

**Comment by fabiospampinato** [comment]
@mikaelmello @MizarcDev @jeff9315 here you can find some builds with the fix implemented, it would be great if you guys could try to reproduce the issue with these:

- Mac (dmg): https://drive.google.com/open?id=1ej2kJE0UlxuBk86KVgcZOYginmPe-AzV 
- Windows (exe): https://drive.google.com/open?id=1vsG0LLd5aAXQ_rhVnTb6HMZQZRGYiaoN

**Comment by MizarcDev** [comment]
@fabiospampinato It appears that this build running on Windows has patched the note override problem. I cannot reproduce the issue using a touch screen, keyboard or mouse.

**Comment by jeff9315** [comment]
OK ... Here's where it stands (on Win 10). 

**Situation 1:** If I edit a note, and while editing it, switch to another note and back, both notes show as being in edit mode (and the file system shows both as having been changed even if the 2nd note was not), but they're both kept which is good news.

**Situation 2:** I use syncthing to synchronize my folders between my android phone and Windows. If I make a change on Android and the synchronize occurs while I am **not** viewing the same file on Windows, everything is OK. However, if I **am** viewing the same file on Windows, the file is deleted. This is consistent. I just tried it about 5 times. Editing my files on Windows and Android is a critical part of my process so I'll probably switch to Typora until the bug is resolved.

As a side note, bugs are one reason why I want my Markup files to be completely portable across products (back and forth). If one product becomes buggy I want to avoid using that product in certain situations but I may want to continue using it in others (using multiple products with the same md file). If you start using (indecipherable) UUIDs, that portability is lost and I can't use your product regardless of whether your have an Android product or not.

**Comment by fabiospampinato** [comment]
>  both notes show as being in edit mode

@jeff9315 How? What visual cue makes you say that? Since there's only one editor visible at a time I'm not sure what you mean with that.

> and the file system shows both as having been changed even if the 2nd note was not

@jeff9315 This shouldn't happen, let's track it [here](https://github.com/notable/notable/issues/641).

>  I use syncthing to synchronize

@jeff9315 let's track this [here](https://github.com/notable/notable/issues/642).

> As a side note, bugs are one reason why I want my Markup files to be completely portable across products (back and forth). If one product becomes buggy I want to avoid using that product in certain situations but I may want to continue using it in others (using multiple products with the same md file). If you start using (indecipherable) UUIDs, that portability is lost and I can't use your product regardless of whether your have an Android product or not.

You definitely don't want your notes to be written in XML or something if you want to edit them with multiple apps at the same time or if you may want to switch to something else in the future.

Regarding UUIDs I don't think it will be necessary to use them in the filenames, putting them in the metadata section should suffice.

**Comment by fabiospampinato** [comment]
I'm closing this as this particular issue seems resolved, thanks everyone for the collaboration.

**Comment by jeff9315** [comment]
> both notes show as being in edit mode

What I meant is that by switching between 2 files, BOTH are immediately shown in the source view without you having to hit the pencil icon.

**Comment by fabiospampinato** [comment]
@jeff9315 both notes' source is being shown at the same time? I'm not sure I'm understanding this right, can you post a screenshot or something?

**Comment by MizarcDev** [comment]
@jeff9315 I may be mistaken, but edit mode being persistent while switching between notes seems to be an intended feature. It only exits edit mode when you intentionally click the edit button or use the shortcut.

**Comment by fabiospampinato** [comment]
@MizarcDev if that's what @jeff9315 means yeah, that's intended, some people want to keep staying in edit mode.
